satpy/multiscene.py

```diff
@@ -69,20 +78,52 @@ def stack(datasets, weights=None, combine_times=True):
     return base
 
 
-def _stack_weighted(datasets, weights, combine_times):
-    """Stack datasets using weights."""
-    weights = set_weights_to_zero_where_invalid(datasets, weights)
+def _stack_blended(datasets, weights, combine_times):
+    """Stack datasets blending overlap using weights."""
+    attrs = combine_metadata(*[x.attrs for x in datasets])
+
+    if combine_times:
+        if 'start_time' in attrs and 'end_time' in attrs:
+            attrs['start_time'], attrs['end_time'] = _get_combined_start_end_times(*[x.attrs for x in datasets])
+
+    overlays = []
+    for weight, overlay in zip(weights, datasets):
+        overlays.append(overlay.fillna(0) * weight)
+
+    base = sum(overlays) / sum(weights, start=1.e-9)
+
+    dims = datasets[0].dims
+    blended_array = xr.DataArray(base, dims=dims, attrs=attrs)
+    return blended_array
 
+
+def _stack_selected(datasets, weights, combine_times, bands):
+    """Stack datasets selecting pixels using weights."""
     indices = da.argmax(da.dstack(weights), axis=-1)
+    if bands:
+        indices = [indices, indices, indices]
+
     attrs = combine_metadata(*[x.attrs for x in datasets])
 
     if combine_times:
         if 'start_time' in attrs and 'end_time' in attrs:
             attrs['start_time'], attrs['end_time'] = _get_combined_start_end_times(*[x.attrs for x in datasets])
 
     dims = datasets[0].dims
-    weighted_array = xr.DataArray(da.choose(indices, datasets), dims=dims, attrs=attrs)
-    return weighted_array
+    coords = datasets[0].coords
+    selected_array = xr.DataArray(da.choose(indices, datasets), dims=dims, coords=coords, attrs=attrs)
+    return selected_array
+
+
+def set_weights_to_zero_where_invalid_red(datasets, weights):
+    """Go through the weights and set to pixel values to zero where corresponding datasets[0] are invalid."""
+    for i, dataset in enumerate(datasets):
+        try:
+            weights[i] = xr.where(dataset[0] == dataset.attrs["_FillValue"], 0, weights[i])
+        except KeyError:
+            weights[i] = xr.where(dataset[0].isnull(), 0, weights[i])
```

Comment:
@lobsiger I'm refactoring this while also writing tests, but noticed that this says `datasets[0]`, should this be `datasets[i]`? Don't worry about fixing it yourself if this is wrong, I'll do it.

Reply:
Oh! This is only looking at the red channel. Why?

Reply:
> Oh! This is only looking at the red channel. Why?

@djhoese this is simply because I was unable to write an accepted function that treated all available bands properly. So I just copied the code @adybbroe wrote for single band cloud data and looked at the RED band only. My (probably wrong) assumption was that either none or all bands would be null or have the _FillValue. At least it seemed to correct OLCI ERR data images. Without this function those images had black lines at the ends (left/right) of the swath of each satellite pass.